### ISSUE TITLE
[Storybook 6.0.0-beta.16] HMR not working properly

### ISSUE THREAD

**Issue by yannbf** [storybookjs/storybook#10949]
**Describe the bug**

There seems to be an issue with Hot module reloading and the latest changes in Storybook. 
Webpack seems to pickup the changes and rebuild accordingly, but it's behaving really weirdly in a lot of different scenarios (click on each gif to open in better resolution):

**1 - Changing arbitrary code** 
It's necessary to refresh in order to see the changes.
![Kapture 2020-05-27 at 20 42 12](https://user-images.githubusercontent.com/1671563/83059723-a55c1e80-a05a-11ea-9f23-1413478d6da4.gif)

 
**2 - Renaming the default title of a story**
It's duplicating the stories in the manager.
![namebug](https://user-images.githubusercontent.com/1671563/83059224-e0118700-a059-11ea-8a53-7a927eaa8107.gif)

 
**3 - Adding/removing stories**
It's not reacting accordingly
![Kapture 2020-05-27 at 20 35 59](https://user-images.githubusercontent.com/1671563/83059203-d38d2e80-a059-11ea-9864-4a4e92efda98.gif)

At the end of the video above I commented all stories, and as a result I got this (render of the button + message saying there was no preview)
![image](https://user-images.githubusercontent.com/1671563/83060313-890cb180-a05b-11ea-95c8-44bc4eee9378.png)

**To Reproduce**
Use the gifs on top as a reference.

**Expected behavior**
HMR should be working correctly, with changes being picked up and reacting accordingly.

**Screenshots**
![image](https://user-images.githubusercontent.com/1671563/83059083-a8a2da80-a059-11ea-90b7-c41f346c0682.png)
![Kapture 2020-05-27 at 20 35 59](https://user-images.githubusercontent.com/1671563/83059203-d38d2e80-a059-11ea-9864-4a4e92efda98.gif)
![namebug](https://user-images.githubusercontent.com/1671563/83059224-e0118700-a059-11ea-8a53-7a927eaa8107.gif)

The issue happened to me using a starter from Vue + TS and the starter from sb init. I will check if it's happening with other frameworks, although I believe it is.

**System:**
**Environment Info:**

  ```
System:
    OS: macOS 10.15.4
    CPU: (8) x64 Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
  Binaries:
    Node: 12.16.1 - /usr/local/bin/node
    Yarn: 1.22.4 - /usr/local/bin/yarn
    npm: 6.13.4 - /usr/local/bin/npm
  Browsers:
    Chrome: 83.0.4103.61
    Firefox: 74.0
    Safari: 13.1
  npmPackages:
    @storybook/addon-actions: ^6.0.0-beta.16 => 6.0.0-beta.16 
    @storybook/addon-links: ^6.0.0-beta.16 => 6.0.0-beta.16 
    @storybook/addons: ^6.0.0-beta.16 => 6.0.0-beta.16 
    @storybook/vue: ^6.0.0-beta.16 => 6.0.0-beta.
```

**Comment by yannbf** [comment]
cc @ndelangen @tmeasday

**Comment by yannbf** [comment]
Can confirm it's behaving similarly in a React project + Beta 16

**Comment by chrisj-skinner** [comment]
I can confirm that point **2 - Renaming the default title of a story** happens in my angular storybook `6.0.0-beta.16` BUT only within `...stories.mdx` files.

`...stories.ts` are not effected.

**Comment by tmeasday** [comment]
Thanks @yannbf

**Comment by ndelangen** [comment]
@tmeasday if you want to work on this together, ping me?

**Comment by tmeasday** [comment]
@ndelangen I'm poking around in this area already. I'll diagnose and bring you in if the fix isn't easy.

**Comment by tmeasday** [comment]
I tried some reproductions here on next and there are some slight differences:

1. **Changing a story:** This is working fine on `next`. Could it be a bug we recently fixed?

2. **Renaming a kind:** This doesn't duplicate, it simply has no effect, more similar to 3.

3. **Adding/removing stories:** This reproduces as described.

In cases 2&3 I noticed that the `SET_STORIES` event is firing and sending over the correct data. So it is a manager issue, likely due to the refs changes @ndelangen -- I might hand this over to you after all.

**Comment by shilman** [comment]
Jeepers creepers!! I just released https://github.com/storybookjs/storybook/releases/tag/v6.0.0-beta.18 containing PR #10981 that references this issue. Upgrade today to try it out!

You can find this prerelease on the `@next` NPM tag.

Closing this issue. Please re-open if you think there's still more to do.

**Comment by eranshmil** [comment]
> Jeepers creepers!! I just released [`v6.0.0-beta.18` (release)](https://github.com/storybookjs/storybook/releases/tag/v6.0.0-beta.18) containing PR #10981 that references this issue. Upgrade today to try it out!
> 
> You can find this prerelease on the `@next` NPM tag.
> 
> Closing this issue. Please re-open if you think there's still more to do.

Using version `6.0.0-beta.19` with an Angular project, I'm having the following issue:
![HMR1](https://user-images.githubusercontent.com/9721664/83350160-d9795d00-a342-11ea-83ba-ba95c30a7c8f.gif)

Before this fix, there was a different issue:
![HMR2](https://user-images.githubusercontent.com/9721664/83350101-5e17ab80-a342-11ea-8661-cad670283170.gif)

**Comment by ndelangen** [comment]
Could you have a reproduction perhaps?

I suspect what's happening is you're writing stories for the same component in multiple story-files.

This messes up the storyStore. If you keep all stories of a component in a single story-file, does the problem still occur?

**Comment by eranshmil** [comment]
> Could you have a reproduction perhaps?
> 
> I suspect what's happening is you're writing stories for the same component in multiple story-files.
> 
> This messes up the storyStore. If you keep all stories of a component in a single story-file, does the problem still occur?
I'm using the same `title` in multiple files, in order to group them under the same parent. I thought it's should be supported.

**Comment by ndelangen** [comment]
@eranshmil we assume every storyFile to have a unique component / title.

When HMR happens we remove all stories of a particular title and then run add the stories again from the file that was HMR'd.

This is why you see the story from the other file being removed.

**Comment by eranshmil** [comment]
> @eranshmil we assume every storyFile to have a unique component / title.
> 
> When HMR happens we remove all stories of a particular title and then run add the stories again from the file that was HMR'd.
> 
> This is why you see the story from the other file being removed.

So it is not possible to group stories under the same parent from different story files?

**Comment by ndelangen** [comment]
If you want to fragment stories of the same group into different files, you could try importing them?

**Comment by shilman** [comment]
Will fix (by explicitly disabling) in #11108
